modules/cart/src/Plugin/Block/CartBlock.php

```diff
@@ -109,11 +110,18 @@ public function blockSubmit($form, FormStateInterface $form_state) {
    *   A render array.
    */
   public function build() {
+    $cachable_metadata = new CacheableMetadata();
+    $cachable_metadata->addCacheContexts(['user', 'session']);
+
     /** @var \Drupal\commerce_order\Entity\OrderInterface[] $carts */
     $carts = $this->cartProvider->getCarts();
     $carts = array_filter($carts, function ($cart) {
       /** @var \Drupal\commerce_order\Entity\OrderInterface $cart */
-      return $cart->hasLineItems();
+      // There is a chance the cart may have converted from a draft order, but
+      // is still in session. Such as just completing check out. So we verify
+      // that the cart is still a cart.
+      // @todo find better place to put this, such as provider::isFinalized
+      return $cart->hasLineItems() && $cart->cart->value;
```

Comment:
Right now when a checkout completes the order ID is still in session, which allows us to support anonymous access to an order. However, the order is no longer a cart but causes it to be returned by the cart provider.

If we adjust the provider's `::getCartIds` or `::getCarts` we'll the same issue we would have if we deleted the ID from the session - no idea if the anonymous user owned the cart.

Right now our main bug is the fact the cart block shows this anonymous order in the cart block on checkout complete (and probably any other page until the session expires.) This check prevents this by validating the returned cart is still a cart.

@bojanz should we commit this with the `@todo` or find fix. Trying to outweigh fix vs. lingering bug